nodejs/aws-infra/experimental/image.ts

```diff
@@ -0,0 +1,206 @@
+// Copyright 2016-2018, Pulumi Corporation.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+import * as aws from "@pulumi/aws";
+import * as docker from "@pulumi/docker";
+import * as pulumi from "@pulumi/pulumi";
+
+import * as utils from "../utils";
+
+import * as mod from ".";
+
+export type ImageEnvironment = Record<string, pulumi.Input<string>>;
+
+export interface IImageProvider {
+    image(name: string, parent: pulumi.Resource): pulumi.Input<string>;
+    environment(name: string, parent: pulumi.Resource): pulumi.Input<aws.ecs.KeyValuePair[]>;
+}
+
+export abstract class ImageProvider implements IImageProvider {
+    public abstract image(name: string, parent: pulumi.Resource): pulumi.Input<string>;
+    public abstract environment(name: string, parent: pulumi.Resource): pulumi.Input<aws.ecs.KeyValuePair[]>;
+
+    /**
+     * Creates a [ContainerImage] given a path to a folder in which a Docker build should be run.
+     */
+    public static fromPath(path: string): IImageProvider {
+        return new AssetImageProvider(path);
+    }
+
+    /**
+     *Creates a [ContainerImage] using the detailed build instructions provided in [build].
+     */
+
+    public static fromDockerBuild(build: docker.DockerBuild): IImageProvider {
+        return new AssetImageProvider(build);
+    }
+
+    /**
+     * Creates a [ContainerImage] given function code to use as the implementation of the container.
+     */
+    public static fromFunction(func: () => void): IImageProvider {
+        return new FunctionImageProvider(func);
+    }
+}
+
+class FunctionImageProvider extends ImageProvider {
+    constructor(private readonly func: () => void) {
+        super();
+    }
+
+    public image(name: string, parent: pulumi.Resource): pulumi.Input<string> {
+        // TODO[pulumi/pulumi-cloud#85]: move this to a Pulumi Docker Hub account.
+        return "lukehoban/nodejsrunner";
+    }
+
+    public environment(name: string, parent: pulumi.Resource): pulumi.Input<aws.ecs.KeyValuePair[]> {
+        const serialized = pulumi.runtime.serializeFunctionAsync(this.func);
+        return serialized.then(value => [{
+            name: "PULUMI_SRC",
+            value: value,
+        }]);
+    }
```

Comment:
things are much cleaner since tehre is no explicit dispatch logic anymore.  Each majorly separate form of image-providing is just in its own type.